lib/sshutils/sftp/sftp.go

```diff
@@ -199,94 +196,85 @@ func (c *Config) initFS(sshClient *ssh.Client, client *sftp.Client) error {
 	}
 
 	return trace.Wrap(c.expandPaths(srcOK, dstOK))
 }
 
 func (c *Config) expandPaths(srcIsRemote, dstIsRemote bool) (err error) {
-	srcHomeRetriever := getLocalHomeDir
-	if srcIsRemote {
-		srcHomeRetriever = c.getHomeDir
-	}
 	for i, srcPath := range c.srcPaths {
-		c.srcPaths[i], err = expandPath(srcPath, srcHomeRetriever)
+		c.srcPaths[i], err = expandPath(srcPath, srcIsRemote, c.getHomeDir)
 		if err != nil {
-			return trace.Wrap(err)
+			return trace.Wrap(err, "error expanding %q", srcPath)
 		}
 	}
 
-	dstHomeRetriever := getLocalHomeDir
-	if dstIsRemote {
-		dstHomeRetriever = c.getHomeDir
-	}
-	c.dstPath, err = expandPath(c.dstPath, dstHomeRetriever)
-
-	return trace.Wrap(err)
+	c.dstPath, err = expandPath(c.dstPath, dstIsRemote, c.getHomeDir)
+	return trace.Wrap(err, "error expanding %q", c.dstPath)
 }
 
-func getLocalHomeDir() (string, error) {
-	u, err := user.Current()
-	if err != nil {
-		return "", trace.Wrap(err)
+func expandPath(pathStr string, isRemotePath bool, getHomeDir homeDirRetriever) (string, error) {
+	pfxLen := homeDirPrefixLen(pathStr)
+	if pfxLen == -1 {
+		return pathStr, nil
 	}
-	return u.HomeDir, nil
-}
 
-func expandPath(pathStr string, getHomeDir homeDirRetriever) (string, error) {
-	if !needsExpansion(pathStr) {
-		return pathStr, nil
+	// if the path is local, expand the path to be rooted at this user's
+	// home dir
+	if !isRemotePath {
+		homeDir, err := getHomeDir()
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+		// this is safe because we verified that all paths are non-empty
+		// in CreateUploadConfig/CreateDownloadConfig
+		return path.Join(homeDir, pathStr[1:]), nil
 	}
 
-	homeDir, err := getHomeDir()
-	if err != nil {
-		return "", trace.Wrap(err)
+	// avoid returning an empty string
+	if pathStr == "~" {
+		return ".", nil
 	}
 
-	// this is safe because we verified that all paths are non-empty
-	// in CreateUploadConfig/CreateDownloadConfig
-	return path.Join(homeDir, pathStr[1:]), nil
+	// if an SFTP path is not absolute, it is assumed to start at the user's
+	// home directory so just strip the prefix and let the SFTP server
+	// figure out the correct remote path
+	return pathStr[pfxLen:], nil
 }
 
-// needsExpansion returns true if path is '~', '~/', or '~\' on Windows
-func needsExpansion(path string) bool {
-	if len(path) == 1 {
-		return path == "~"
+// homeDirPrefixLen returns the length of a set of characters that
+// indicates the user wants the path to begin with a user's home
+// directory, or -1 if no such prefix exists.
+func homeDirPrefixLen(path string) int {
+	if len(path) == 1 && path == "~" {
+		return 1
+	}
+	if strings.HasPrefix(path, "~/") {
+		return 2
 	}
-
 	// allow '~\' or '~/' on Windows since '\' is the canonical path
 	// separator but some users may use '/' instead
 	if runtime.GOOS == "windows" && strings.HasPrefix(path, `~\`) {
-		return true
+		return 2
 	}
-	return strings.HasPrefix(path, "~/")
+	if path[0] == '~' {
+		return 1
+	}
```

Comment:
@capnspacehook What does OpenSSH do in this situation?

Reply:
You're right, changed to only expand `~user` paths when they are local and return an error otherwise. Expanding remote paths in this way requires SFTP server support, I could easily implement it in a Teleport-specific way but I'm leaning towards not alienating OpenSSH `scp` users.

I might try and open an upstream PR that adds the ability for our SFTP server to do this, but it's complicated as OpenSSH `scp` will try and use a OpenSSH specific request to resolve home dirs, while most other SFTP clients will use a different standardized request. I may have to implement both, but idk if upstream maintainers will want that.

Reply:
I'll tackle that in another PR though, and we already have an [issue to track it](https://github.com/gravitational/teleport/issues/22886) so I'm going to mark this resolved now.